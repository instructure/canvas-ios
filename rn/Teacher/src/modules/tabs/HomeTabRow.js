//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

// @flow

import React from 'react'
import 'react-native'
import { SafeAreaView } from 'react-native'

import i18n from 'format-message'
import FeatureRow from '../../common/components/rows/FeatureRow'

type Props = {
  tab: Tab,
  defaultView: ?string,
  onPress: Function,
  selected: ?boolean,
}

export default class CourseDetailsHomeTab extends React.Component<Props, any> {
  onPress = () => {
    const tab = this.props.tab
    this.props.onPress(tab)
  }

  subtitle = (): ?string => {
    switch (this.props.defaultView) {
      case 'assignments': return i18n('Assignments')
      case 'feed': return i18n('Recent Activity')
      case 'wiki': return i18n('Front Page')
      case 'modules': return i18n('Course Modules')
      case 'syllabus': return i18n('Syllabus')
      default: return null
    }
  }

  // Notice that CourseDetailsTab.js has a safe area around it, but this doesn't
  // There is a rendering issue that causes the containing scroll view to freak out.
  // We decided that because this tab is only shown in the student app and the student app doesn't support landscape, we don't
  // Need to have the safe area here, and that consequently fixed the layout issue.
  render () {
    const tab = this.props.tab
    return (
      <SafeAreaView>
        <FeatureRow
          title={tab.label}
          subtitle={this.subtitle()}
          onPress={this.onPress}
          disclosureIndicator
          selected={this.props.selected}
          testID={`courses-details.${tab.id}-cell`}
        />
      </SafeAreaView>
    )
  }
}
