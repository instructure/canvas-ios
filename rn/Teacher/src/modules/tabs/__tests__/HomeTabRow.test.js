//
// This file is part of Canvas.
// Copyright (C) 2018-present  Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.
//

/* eslint-disable flowtype/require-valid-file-annotation */

import { shallow } from 'enzyme'
import React from 'react'
import HomeTabRow from '../HomeTabRow'

const template = {
  ...require('../../../__templates__/tab'),
  ...require('../../../__templates__/course'),
}

const defaultProps = {
  tab: template.tab(),
  course: template.course({ 'default_view': 'assignments' }),
  courseColor: 'white',
  onPress: () => {},
}

describe('HomeTabRow', () => {
  it('renders correctly', () => {
    const tree = shallow(<HomeTabRow {...defaultProps} />)
    expect(tree).toMatchSnapshot()
  })

  it('renders correctly for each type', () => {
    const types = ['assignments', 'feed', 'wiki', 'modules', 'syllabus', 'whatisthis']
    types.forEach((type) => {
      const course = template.course({ 'default_view': type })
      const tree = shallow(<HomeTabRow {...defaultProps} course={course} />)
      expect(tree).toMatchSnapshot()
    })
  })

  it('can be selected', () => {
    const tree = shallow(<HomeTabRow {...defaultProps} selected />)
    expect(tree).toMatchSnapshot()
  })

  it('can be pressed', () => {
    const onPressed = jest.fn()
    const props = {
      ...defaultProps,
      onPress: onPressed,
    }

    const tree = shallow(<HomeTabRow {...props} />)
    expect(tree).toMatchSnapshot()
    tree.find('FeatureRow').simulate('press')
    expect(onPressed).toHaveBeenCalledTimes(1)
  })
})
