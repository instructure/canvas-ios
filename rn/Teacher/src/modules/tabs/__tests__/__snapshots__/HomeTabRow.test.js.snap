// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeTabRow can be pressed 1`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow can be selected 1`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    selected={true}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly 1`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly for each type 1`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly for each type 2`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly for each type 3`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly for each type 4`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly for each type 5`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;

exports[`HomeTabRow renders correctly for each type 6`] = `
<ForwardRef(SafeAreaView)>
  <FeatureRow
    disclosureIndicator={true}
    onPress={[Function]}
    subtitle={null}
    testID="courses-details.assignments-cell"
    title="Home"
  />
</ForwardRef(SafeAreaView)>
`;
